src/core/public/context/context.ts

```diff
@@ -0,0 +1,281 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { flatten } from 'lodash';
+import { PluginName } from '../../server';
+import { pick } from '../../utils';
+import { CoreId } from '../core_system';
+
+/**
+ * A function that returns a context value for a specific key of given context type.
+ *
+ * @remarks
+ * This function will be called each time a new context is built for a handler invocation.
+ *
+ * @param context - A partial context object containing only the keys for values provided by plugin dependencies
+ * @param rest - Additional parameters provided by the service owner of this context
+ * @returns The context value associated with this key. May also return a Promise.
+ *
+ * @public
+ */
+export type ContextProvider<
+  TContext extends {},
```

Comment:
`extends Record<string, any>` or `extends object`? for the same reasons as https://github.com/elastic/kibana/pull/40846